A method to alleviate the impact of object searching on frame rate.

2822867502 opened the pull request:
针对该issue #14 的解决方案之一：
使用Unity的新API `FindObjectsOfType`并禁用排序（`FindObjectsSortMode.None`），以此减弱去码对帧率影响
再次向作者表达感谢

One of the solutions for issue #14:
Use Unity's new API `FindObjectsOfType` with sorting disabled (`FindObjectsSortMode.None`), in order to reduce the impact of demosaicing on frame rate.
Once again, I'd like to express my gratitude to the author.

2822867502 committed:
Use the new API 'FindObjectsByType' to alleviate the impact of object searching on frame rate.

ManlyMarco reviewed (commented):
This is a good fix, but the issue is that a lot of games do not support this API, since it was added in Unity 2022.2 https://docs.unity3d.com/2022.2/Documentation/ScriptReference/Object.FindObjectsByType.html

One way to resolve this would be to build two versions of the plugin, one for Unity 2022.2+ and one for all versions.
I'm not sure if there are many IL2CPP games that don't have this API but it is possible.

A more practical solution would be to check if the new API is available with reflection and only call it then. It's a little slower but avoids having two separate builds.

2822867502 commented:
Thank you for your support.

I think what you said makes sense — the Unity version is indeed an important factor to consider.

I took another look at the decompiled source code of the Unity engine, and found that the object-finding process itself isn't particularly time-consuming. So I started thinking — maybe we can approach this from a different angle.

For example, what if we check whether the number of Renderer objects in the current frame is the same as in the previous frame? If they are the same, we could assume that no new Renderer objects were created or destroyed, and thus skip the more expensive processing steps most of the time.

I realize this might seem like a bold assumption, but I believe the chance of having exactly the same number of Renderer objects being destroyed and created within a single frame is so small that it's practically negligible.

Another idea is to process the data at intervals across multiple frames. Since the human eye can't easily perceive changes over very short periods of time, perhaps we could reduce the processing rate in exchange for better performance. For instance, instead of processing everything in one frame, we could spread it across two or more frames, effectively doubling the available processing time. This interval could also be made configurable (e.g., via a .cfg file), allowing players to adjust it based on their hardware.

What do you think about these ideas?

Translated by Qwen (Tongyi Qianwen)

2822867502 committed:
Compatibility improvements have been made.
Note: Due to development environment and resource limitations, only the .NET 6 version was compiled, and testing was only performed on versions 2022.2.2 and above.

2822867502 commented:
I followed your instructions and added version checking and reflection calls. However, as I mentioned in #14, I am not a professional Unity developer, and I only have a .NET 6 compilation environment. Therefore, I only tested the .NET 6 version. I made changes to the other versions as well, but I did not compile or test them (in fact, my IDE displays them in a complete mess for the other versions, not even auto-formatting works).

I'm very sorry—I wanted to help, but ended up causing you more trouble, requiring you to review and test my code. I will strive to improve my Unity development skills and continue contributing to the community.

By the way, during the compilation of the .NET 6 version, there was a missing `{}` issue. After I changed 
```
namespace DumbRendererDemosaic;

......
```
 to 
```
namespace DumbRendererDemosaic{
......
}
```
,the compilation worked normally—though I’m not sure whether the problem was in the code or in my compiler. In any case, I did not make any changes to this part in my final submission.

Translated by Qwen

ManlyMarco commented on DumbRendererDemosaic/DumbRendererDemosaic.cs:
```diff
@@ -11,20 +14,45 @@ namespace DumbRendererDemosaic
         {
             MozaicTools.InitSetting(Config);
         }
+        private static bool IsUp2022_2 = DumbRendererDemosaicPlugin.UnityVersion >= new Version("2022.2");
```
This way of version checking is not reliable because some builds of Unity before 2022.2 do include this API. It's much simpler and more reliable to just try getting the method with reflection, and if it doesn't exist fall back to the commonly available FindObjectsOfType.

ManlyMarco commented on DumbRendererDemosaic/DumbRendererDemosaic.cs:
```diff
@@ -11,20 +14,45 @@ namespace DumbRendererDemosaic
+        Il2CppReferenceArray<UnityEngine.Object> array =
+        				(Il2CppReferenceArray<UnityEngine.Object>)(IsUp2022_2 ?
+        				FindObject.Invoke(null,new object[] {Il2CppType.Of<Renderer>(),FindObjectsSortMode.None}) :
+        				FindObject.Invoke(null,new object[] {Il2CppType.Of<Renderer>(),false}));
```
Rather than checking the version every time to select how to Invoke, it would be simpler to store this as a delegate in Start (e.g. `_delegate = new Func<Renderer[]>(()=> FindObject.Invoke(null,new object[] {Il2CppType.Of<Renderer>(),FindObjectsSortMode.None}))`).

ManlyMarco reviewed (changes requested):
Thanks for the update. It's good to see that there is interest in this project.

The idea is good, it just needs some optimization. Rather than adding all of the new fields it can be done by adding one field `Func<Renderer[]> _findObjDelegate;` and assigning it from Start.

If the foreach loop is the slow part of the code then it can be put into a coroutine to split it across frames to lessen the impact. `IEnumerator Start()` can be used in that case to simplify the code (make sure to use the interop IEnumerator for IL2CPP). It would be best to do this in a separate PR however.

The namespace issue is likely because you're using an older version of VisualStudio or another IDE that doesn't support new language features.

2822867502 commented:
Okay, I will modify the code as soon as possible to achieve the best results.
I sincerely apologize for keeping you waiting so long. I've been trying to fix my development environment, but due to various reasons (for example, running out of disk space, the Visual Studio download being too slow, and missing dependency libraries for Rider), it has taken me much longer than expected and I haven't been successful yet. Thank you for your understanding and support.

2822867502 committed:
Revert "Compatibility improvements have been made."

This reverts commit 1f554928b7f5d4e1370bd42ce9a753b757cfb854.

2822867502 committed:
Revert "Use the new API 'FindObjectsByType' to alleviate the impact of object searching on frame rate."

This reverts commit 43c1de9781d425ca4ad74334756267488596ad30.

2822867502 committed:
Use coroutines to handle time-consuming traversal operations instead of finding Objects

2822867502 commented:
Thank you for your patience. After modifying my code and conducting tests, I discovered that the most time-consuming part was the traversal process. Therefore, I refactored the traversal using coroutines to process exactly 100 Renderers per frame. I experimented with other values but found that processing between 100 to 200 Renderers per frame yielded optimal results (in fact, using 128 combined with bitwise left shift operators would be more efficient, but at the cost of reduced code maintainability). Below is partial recorded test data:

| About 3000 Renderers | Legacy API Calls | New API Calls | Reflection Calls |
|----------------------|------------------|---------------|------------------|
| **Finding Objects**  | 2ms             | ~1ms          | 2ms             |
| **Select Operation** | 0ms             | 0ms           | 0ms             |
| **Traversal**        | 305ms               | 305ms         | 305ms           |

Translated by DeepSeek

2822867502 commented:
I sincerely apologize that the automatic formatting from the IDE has made the git diff display somewhat messy.

ManlyMarco reviewed (approved):
Looks good to me, thanks for the PR!

ManlyMarco commented:
Made a release with these changes https://github.com/ManlyMarco/UniversalUnityDemosaics/releases/tag/v1.7

2822867502 commented:
Thank you equally for your patient guidance. You've helped transform me from someone who knew Java but was completely unfamiliar with C# into a developer who now understands some C# features. Although I'm still a programming novice, I believe that someday in the future I'll become as skilled as you are. Once again, thank you for your patience and support.
